### DISCUSSION TITLE
Minestom agent implementation & Custom plugin folder location

### DISCUSSION THREAD

**Discussion by TropicalShadow** [jeremylvln/Shulker#362] [Ideas]
Hi, I am looking at integrating [Minestom](https://minestom.net/) \([GitHub](https://github.com/Minestom/Minestom)\) into this project. 

I have attempted implementing these changes in my fork [BridgeSplash Fork](https://github.com/BridgeSplash/Shulker) 

here is a rough list of features/fixes I added to my fork:

- Custom Plugin Folder Location [Commit](https://github.com/BridgeSplash/Shulker/commit/5f602b4d164b23520126a972564f6423538a1dbb)
- Custom Server Interface, edited Logging for Servers [Commit](https://github.com/BridgeSplash/Shulker/commit/1e1add3226d82744331bdb67ff8616a62d7bb2ae)
- Fixed .gitignore for gradle auto-import [Commit](https://github.com/BridgeSplash/Shulker/commit/d57c7c29f2ad0af3f15256b660455f61bc9d6a55)

all commits can be found [here](https://github.com/BridgeSplash/Shulker/commits/main/)

_Note:_ I still need to test that the changes work with no issue

Sorry if this is posted to the wrong section, was unsure where to post.
hope this can help, and I am happy to tweak or change it if needed.
Thank you, 
Tropical.

**Comment by jeremylvln** [comment]
Hi!

Thanks a lot for opening the discussion, it's clearly the appropriate place!

_First, a HUGE thank you for the Gradle auto-import fix in the `.gitignore`. I didn't see that I had to un-comment a part from the template I used, I've already pushed it on the main branch._

About Minestom, I didn't know it existed before and that's a tough subject for me. I'm not closed in adding support for other platforms (at one point I will have to do it for Forge and Fabric servers), but they are not _niche_ platforms and I know adding support for them will benefit a lot of people.

I think the point here is to have a good balance between time investment in supporting a _niche_ platform, and actually the reward being the number of people interested in it.

It's not only adding a platform plugin, it's also configuring the operator to recognize Minestom as a platform (which is not hard, my point is that it impacts multiple parts of the codebase).

As you already did most of the job, I'm absolutely not closed into merging this. But we may also understand that, if I had to prioritize, maintaining it will not be my top priority.

Let me cherry-pick your commits, I'll do the operator part and open a PR really soon.

Thanks!

**Reply by TropicalShadow** [reply]
I appreciate the response! 
ideally if the feature to change the plugins folder location would be implemented, that would be perfectly fine for running Minestom.

**Reply by jeremylvln** [reply]
Can you explain more in depth the plugin folder location issue? Does it have another name in Minestom? Always the same?
Just to know if I have to provide a configuration field or if I can hardcode a path if the platform is Minestom.

**Reply by TropicalShadow** [reply]
The "plugins" folder on Minestom is called "extensions".

- Hardcoding a path may work, but if you look at my commit I add the possibility to configure the plugins folder location from the config [here](https://github.com/BridgeSplash/Shulker/commit/5f602b4d164b23520126a972564f6423538a1dbb), this would be much more preferred as it can be used later on with any other fork or alternative Minecraft server software which may not use "plugins" as their addon folder i.e "mods" & "extensions"

- I would also request you look at adding an interface Logging class like I have in this [commit](https://github.com/BridgeSplash/Shulker/commit/1e1add3226d82744331bdb67ff8616a62d7bb2ae) - [file](https://github.com/BridgeSplash/Shulker/blob/1e1add3226d82744331bdb67ff8616a62d7bb2ae/packages/shulker-server-agent/src/common/kotlin/io/shulkermc/serveragent/logger/AgentLoggerInterface.kt)
that will allow for custom agents to use alternative logging (like Minestom)

**Reply by jeremylvln** [reply]
I looked into it earler today and I don't think I want to have support for Minestom for now. There are two reasons that are really no-go to me:

1. The project is evolving a lot quite recently, a lot of code is changed due to the upstream of the "community edition" fork. This includes the pure deletion of the extensions API (it's still available as a separate library, but it shows the intent to move back from this). I can't justify the investment in time for something that is changing that much, especially because I don't know the tool. I would love to learn more about it, but I don't want to have to redo this multiple times, so I would prefer to wait for things to be more mature
2. There are no stable Maven publishing, releases on Jitpack are ephemeral, this is not acceptable. For Minestom to be supported widely in more tools (and that's also the output of LuckPerms), they will need to publish proper releases on a proper Maven repository (and I'm not speaking of snapshot releases)

Let's keep this discussion open, so newcomer could learn that Minestom support is evaluated. Do not hesitate to get back to me here if things has improved for the community, I would love to finish the implementation properly once these issues are fixed.

Thanks!

**Comment by samuelh2005** [comment]
Now current Minestom no longer have any such thing as extensions, since Minestom CE have been merged back into orignal Minestom.

Minestom has now been redesigned into a generic Java API which means it no longer loads things on the user's behalf. - This makes it harder to load Shulker into Minestom servers, since Shulker must be a dependency of these server, however other libraries like [MinestomPVP](https://github.com/TogAr2/MinestomPvP) already do this.

Also Minestom is now [fully independent of Jitpack](https://github.com/Minestom/Minestom/issues/1987). 

There is also a new Minestom tutorial series on YouTube by Kody Simpson which will make Minestom more popular then it is now.

From my experience with Minestom their API is very simple but powerful. I don't see Shulker needing to rely on most of the API they've made.

**Reply by samuelh2005** [reply]
In the meantime, I've bundled the Shulker agent into my Minestom server, implemented the agent's interfaces and then made my server read the standard `server.properties` and `paper-global.yml` file. This means my Minestom configs would be compatible with Paper's (which is needed for Shulker).

Then I set the [customJar](https://shulker.jeremylvln.fr/latest/guide/recipes/using-custom-server-jar.html) of my server's YAML to point one of my Minestom server.

In theory, this would *have worked without any changes to Shulker's backend* **(which would mean no maintenance is needed)**, however there is a problem:

> [init] Starting the Minecraft server...
Error: LinkageError occurred while loading main class uk.minersonline.games.Main
        java.lang.UnsupportedClassVersionError: uk/minersonline/games/Main has been compiled by a more recent version of the Java Runtime (class file version 65.0), this version of the Java Runtime only recognizes class file versions up to 61.0
        
 It looks like Shulker does not use Java 21 for running it's server pods. Minestom (and newer versions of Minecraft) require Java 21 to run, but here Java 17 (61.0) is used. ***Is there a way to change the Java version?***

**Reply by jeremylvln** [reply]
Would you mind sharing your setup to make Shulker work with Minestom? I found that libraries often provides an `.init()` method that configures everything, it's totally something I can do here. Then, Shulker will not download any agent plugin

**Reply by samuelh2005** [reply]
Yes I can, it requires some refactoring first. I will publish the repo once that has been done.

**Reply by samuelh2005** [reply]
Update: I have finished my refactoring and have published the [Shulker for Minestom](https://github.com/miners-online/shulker-for-minestom) repo.

**Reply by jeremylvln** [reply]
I can do some work to make this a first-party Shulker component (i.e. integrating it in the monorepo). It would help anyone creating a Minestom server without too many hacks. To avoid "reinventing the wheel", I would prefer cherry-picking some code from your repository. Would this be an issue for you?

**Reply by samuelh2005** [reply]
That great.

> Would this be an issue for you?

No, do whatever you need to do to get it working :)
